smol/capp/generate/groundstate/upper_bound/constraints.py

```diff
@@ -0,0 +1,244 @@
+"""Get constraints on variables from a processor."""
+from numbers import Number
+from typing import List, Tuple, Union
+
+import cvxpy as cp
+import numpy as np
+from cvxpy import Expression
+from cvxpy.constraints.constraint import Constraint
+from pymatgen.core import Structure
+
+from smol.capp.generate.groundstate.upper_bound.indices import (
+    get_variable_indices_for_each_composition_component,
+)
+from smol.moca.composition.space import CompositionSpace
+from smol.moca.ensemble import Sublattice
+
+__author__ = "Fengyu Xie"
+
+
+def get_upper_bound_normalization_constraints(
```

Comment:
Ok. But note that in the lower-bound problem, since periodicity is no longer maintained, one needs an additional argument to specify the (i,j,k,p,s) coordinates (for this, refer to Weixuan's formulation in his paper) of each site. This would change the inputs heads of constraint functions. We can consider that later if we ever implement a lower-bound solver.

However, in the meantime I would like to move solver.py under upper_bound module and change the name of GroundStateSolver into UpperBoundSolver, because the solver defined in solver.py is now specific to the ground-state problem.

Reply:
you mean specific to the "upper bound ground state problem"?  I think its still ok and probably would recommend to keep the name as `GroundStateSolver` and mention the details of the upper bound and link to Wenxuan's paper in the docstrings and documentation.

Reply:
I recommend keeping the current implementation because it will allow us to implement the LowerBoundSolver in the future.